core/core.cc

```diff
@@ -346,6 +370,11 @@ void TaskGraph::execute_point(long timestep, long point,
 
   // Execute kernel
   Kernel k(kernel);
+
+  //-- add by Yuankun, Init kernel parameters
+  k.scratch_ptr = scratch_ptr;
+  k.scratch_bytes_per_task = scratch_bytes_per_task;
+  //-- add by Yuankun
   k.execute();
```

Comment:
Could we make `Kernel::execute` takes scratch_ptr and bytes as direct arguments? (Note, for backwards compatibility, we have to keep 2 versions, `Kernel::execute()` and `Kernel::execute(char *, size_t)`, but the first can just call the second with NULL, 0, like we did up above.